### ISSUE TITLE
Distinct with a subset of columns, i.e. DISTINCT ON, DropDuplicates feature

### ISSUE THREAD

**Issue by MisterKloudy** [Eventual-Inc/Daft#3181]
### Is your feature request related to a problem?

Similar to https://github.com/Eventual-Inc/Daft/issues/3082

It would be good to implement a version of distinct which only takes in a subset of the columns and provides a way to select how the rest of the duplicate rows should be dropped.

### Describe the solution you'd like

I'd like to deduplicate my data on a certain column (e.g. some_id) or subset of columns and deterministically get a dataframe with unique 'some_id' accompanied by the same data every time I run the same operation. It would not be great if a different row is kept each time. The other columns have different data so the dataframe-level distinct would not work.

### Additional Context

Pandas: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html which supports selecting first, last, or dropping all duplicates

### Would you like to implement a fix?

No

**Comment by jaychia** [comment]
It's actually fairly tricky to ensure determinism in a distributed setting for the ordering, and we might need maybe some kind of sort ordering parameter to tell us how to grab the "first/last".

I'm guessing this would be something like a `df.groupby("keys").deduplicate(order_by="sort_key", keep="first")`?

**Comment by jaychia** [comment]
cc @universalmind303 also for any thoughts on how this would look like in SQL so we can maybe ensure a more consistent API

**Comment by MisterKloudy** [comment]
That's true that it could be tricky. Perhaps it would be better to split the work into different features:

1. Dropduplicates on subset of columns without guaranteeing order (similar to spark's, groupby any>
2. Dropduplicates on subset of columns which always gets first row (groupby first) <if it's not default to keep the ordering during the groupby then this would be the same as the first>
3. Dropduplicates on subset of columns which implements a predefined sort (groupby, sort, first)
4. Dropduplicates on subset of columns which implements a custom sorting order (groupby, mapgroups sort, first)

**Comment by srilman** [comment]
@MisterKloudy, we just merged a PR to add the option (1). We're going to create follow-up issues for the next steps.

**Comment by MisterKloudy** [comment]
Cool! Thank you! Looking forward to it!
